modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java

```diff
@@ -0,0 +1,567 @@
+            compFut.add(fut);
+
+            futMap.put(partId, fut);
+        }
+
+        /**
+         * @return Compound future, that will be completed when all partitions scanned.
+         */
+        public IgniteInternalFuture<Void> initialize() {
+            return compFut.markInitialized();
+        }
+
+        /**
+         * @return Future that ends after all partitions are done and a checkpoint is finished.
+         */
+        public IgniteInternalFuture<Void> finishFuture() {
+            return cpFut;
+        }
+
+        /**
+         * Finish reencryption future.
+         *
+         * @return {@code True} if the future was finished by this call.
+         */
+        public boolean finish() {
+            return cpFut.onDone(compFut.result());
+        }
+
+        /**
+         * Stop reencryption of the specified partition.
+         *
+         * @param partId Partition ID.
+         * @return {@code True} if reencryption was cancelled.
+         * @throws IgniteCheckedException If failed.
+         */
+        public boolean cancel(int partId) throws IgniteCheckedException {
+            IgniteInternalFuture<Void> fut = futMap.get(partId);
+
+            if (fut == null)
+                return false;
+
+            return fut.cancel();
+        }
+
+        /**
+         * @param t Throwable.
+         */
+        public void fail(Throwable t) {
+            cpFut.onDone(t);
+        }
+    }
+
+    /**
+     * Page store scanning task.
+     */
+    private class PageStoreScanTask extends GridFutureAdapter<Void> implements Runnable {
+        /** Partiion ID. */
+        private final int partId;
+
+        /** Cache group scan context. */
+        private final GroupScanContext scanCtx;
+
+        /**
+         * @param scanCtx Cache group scanning context.
+         * @param partId Partition ID.
+         */
+        public PageStoreScanTask(GroupScanContext scanCtx, int partId) {
+            this.scanCtx = scanCtx;
+            this.partId = partId;
+        }
+
+        /** {@inheritDoc} */
+        @Override public synchronized boolean cancel() throws IgniteCheckedException {
+            return onDone();
+        }
+
+        /** {@inheritDoc} */
+        @Override public void run() {
+            try {
+                int grpId = scanCtx.groupId();
+
+                CacheGroupContext grp = ctx.cache().cacheGroup(grpId);
+
+                if (grp == null) {
+                    onDone();
+
+                    return;
+                }
+
+                PageMemoryEx pageMem = (PageMemoryEx)grp.dataRegion().pageMemory();
+                long metaPageId = pageMem.partitionMetaPageId(grpId, partId);
+                long state = ctx.encryption().getEncryptionState(grpId, partId);
+                int batchSize = encrCfg.getReencryptionBatchSize();
+
+                int off = (int)(state >> Integer.SIZE);
+                int cnt = (int)state;
```

Comment:
Logic for extraction state is located at different classes. Perhaps there should be some helper class or helper methods in one class to extract and generate state out of offset and pages count.

Reply:
Added utility class ReencryptStateUtils for this